feat: Add Kafka Connect Cloud Bigtable sink connector

prawilny opened the pull request:
This PR adds Kafka Connect sink connector.

The code is to land in a different repository, but the repository hasn't been created yet, so we bring the code for early review here.
The fact that it's targetting another repo is the reason of the following:
- modification of Github Actions CI so that it executes the sink's tests instead of adding it as a new maven submodule and plugging it into existing CI
- the code is not a submodule in the root pom.xml nor does it use any information from outside of the directory
- maven plugins' config is duplicated
- there is another copy of the license in the directory

Things yet to be done (in future PRs):
- Logical types support
- More comprehensive integration tests
    - they might include more detailed compatibility checks against the Confluent sink
- Use of kokoro in CI

prawilny committed:
feat: Kafka Connect Cloud BigTable Sink

brandtnewton commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapperTest.java:
```diff
@@ -0,0 +1,835 @@
+
+    Schema kafkaConnectInnerSchema =
+        SchemaBuilder.struct()
+            .field(innerFieldStringName, Schema.STRING_SCHEMA)
```
nit: the "inner" prefix is redundant and made me think there were outer fields

prawilny replied:
Done.
Sorry, I must've left it after a refactor.

brandtnewton commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapperTest.java:
```diff
@@ -0,0 +1,835 @@
+    Struct kafkaConnectStruct = new Struct(kafkaConnectSchema);
+    kafkaConnectStruct.put(nullableFieldName, nullableFieldValue);
+    kafkaConnectStruct.put(requiredFieldName, requiredFieldValue);
+    assertThrows(DataException.class, () -> calculateKey(List.of(), DELIMITER, kafkaConnectStruct));
```
nit: Please add a comment for why this exception is expected

prawilny replied:
Done.

brandtnewton commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapperTest.java:
```diff
@@ -0,0 +1,835 @@
+
+    Struct kafkaConnectStruct = new Struct(kafkaConnectSchema);
+    kafkaConnectStruct.put(nullableFieldName, null);
+    assertThrows(DataException.class, () -> calculateKey(List.of(), DELIMITER, kafkaConnectStruct));
```
nit: Please add a comment for why this exception is expected

prawilny replied:
Done.

brandtnewton reviewed (changes requested):
This is shaping up nicely! All my comments are pretty minor. I did not review the Integration tests yet. I'll get back to you soon on how to handle logical types.

prawilny committed:
Elaborate on error handling in awaitResourceCreationAndHandleInvalidInputErrors()

prawilny committed:
rename createdColumnFamilyFuturesAndRecords

prawilny committed:
Tweak NullValueMode docstring

prawilny committed:
Rename test cases valdiating multiple config parameters

prawilny committed:
Add Map ordering disclaimer

prawilny committed:
Get rid of interpolationish suffix from the record topic's name in testGetTableName

prawilny committed:
Remove mocks from testCreateRecordMutationDataNonemptyKey

prawilny committed:
remove leftover empty test

prawilny committed:
Add ALREADY_EXISTS to the list of tested codes in SchemaApiExceptionsTest

prawilny committed:
remove a wild semicolon

prawilny committed:
Split primitive test cases in KeyMapperTest

prawilny committed:
clean up testDifferentStructMappings

prawilny committed:
Explain why exceptions are expected to be thrown

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/autocreate/BigtableSchemaManager.java:
```diff
@@ -0,0 +1,470 @@
+   *       invalid, it is assumed that input {@link SinkRecord SinkRecord(s)} map to invalid values,
+   *       so all the {@link SinkRecord SinkRecord(s)} needing the resource whose creation failed
+   *       are returned.
+   *   <li>Other resource creation errors are logged.
```
Done. Is it appropriate now?

brandtnewton replied:
Looks good, thanks

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/autocreate/BigtableSchemaManager.java:
```diff
@@ -0,0 +1,470 @@
+      Map<Fut, ResourceAndRecords<Id>> createdColumnFamilyFuturesAndRecords,
+      String errorMessageTemplate) {
+    Set<SinkRecord> dataErrors = new HashSet<>();
+    createdColumnFamilyFuturesAndRecords.forEach(
```
Done.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/mapping/ValueMapper.java:
```diff
@@ -0,0 +1,245 @@
+      for (Map.Entry<Object, Object> field : getChildren(rootKafkaValue)) {
+        String kafkaFieldName = field.getKey().toString();
+        Object kafkaFieldValue = field.getValue();
+        if (kafkaFieldValue == null && nullMode == NullValueMode.IGNORE) {
```
I think it should apply to all the values (including the nested ones).
Rationale:
- we need `NullValueMode.WRITE` to behave like the Confluent's sink and it treats it (within values) as if it were empty byte arrays on all the nesting levels.
  - We got Gary's approval for introduction of this configuration arguing that `null`s causing deletes by default would be a footgun for users migrating from the Confluent's sink. Do you agree or should we rethink this idea?
- we need `NullValue.DELETE` to behave as described in the design doc, which means deletion on all the levels.

Alternatively, we could introduce per-nesting-level configuration, but it seems excessive to me. What do you think?

I tweaked the docstring a bit to make it a bit clearer that this config affects more than only root `null`.

brandtnewton replied:
That makes sense. Let's keep it the way you have it

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/mapping/ValueMapper.java:
```diff
@@ -0,0 +1,245 @@
+            ByteString kafkaSubfieldName =
+                ByteString.copyFrom(subfield.getKey().toString().getBytes(StandardCharsets.UTF_8));
+            Object kafkaSubfieldValue = subfield.getValue();
+            if (kafkaSubfieldValue == null && nullMode == NullValueMode.IGNORE) {
```
I think it should apply to all the values (including the nested ones).
Rationale:
- we need `NullValueMode.WRITE` to behave like the Confluent's sink and it treats it (within values) as if it were empty byte arrays on all the nesting levels.
  - We got Gary's approval for introduction of this configuration arguing that `null`s causing deletes by default would be a footgun for users migrating from the Confluent's sink. Do you agree or should we rethink this idea?
- we need `NullValue.DELETE` to behave as described in the design doc, which means deletion on all the levels.

Alternatively, we could introduce per-nesting-level configuration, but it seems excessive to me. What do you think?

I tweaked the docstring a bit to make it a bit clearer that this config affects more than only root `null`.

brandtnewton replied:
That makes sense. Let's keep it the way you have it

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/mapping/ValueMapper.java:
```diff
@@ -0,0 +1,245 @@
+        }
+      }
+    } else {
+      if (defaultColumnFamily != null && defaultColumnQualifier != null) {
```
FYI we talked with Gary and he approved this additional config.

brandtnewton replied:
Great! Sounds good.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/config/BigtableSinkConfigTest.java:
```diff
@@ -0,0 +1,157 @@
+  }
+
+  @Test
+  public void testMultipleValuesValidationInsert() {
```
Done.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/autocreate/BigtableSchemaManager.java:
```diff
@@ -0,0 +1,470 @@
+          List<SinkRecord> sinkRecords = resourceAndRecords.getRecords();
+          try {
+            fut.get();
+          } catch (ExecutionException | InterruptedException e) {
```
As far as I understand (and hopefully integration tests in the future will show), `AlreadyExistsException` is wrapped in an `ExecutionException` and it does go through the whole code path up to `isStatusCodeCausedByInputError()`, but there the logic decides that it is not an exception caused by input error.
It's intentionally handled only with logging. We group records for which we ensure that tables/column families exist into three groups:
- the ones for which the resources exist,
- the ones for which the resources do not exist and the error was caused due to invalid input (we expect that they map to an invalid table/column family, so that there's no point checking if something else created them concurrently) - these are the result of this function,
- the ones for which the resources do not exist and the error was caused by something else than invalid input, be it a race you mentioned, a network error between Bigtable created the resource and our sink's client received the response or something else I cannot think of at the moment.

Also, we split the error so that the user can differentiate retriable and nonretriable errors if they end up in the DLQ - apart from that, the handling of the latter two groups is the same.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapper.java:
```diff
@@ -0,0 +1,262 @@
+      Map<?, ?> map = (Map<?, ?>) keyElement;
+      return concatenateByteArrays(
+          "{",
+          map.entrySet().stream()
```
Nice catch, I missed it.
It is as deterministic as the configured converter makes it.
I think (but I am not sure, I didn't write it down) I remember that in practice the Confluent's sink sorts the values by key, but I'd expect it to be like this due to it calling `Map#toString()` and its use of some iterator over the map, which turned out to be sorted during my tests.

I first thought that sorting this map would have only positive consequences (predicatability of output, mainly), but then I realized that we don't really know the types the converter passes to us, so we'd have to write some possibly tricky code to handle `Object`s of incomparable types to avoid throwing an exception that wouldn't be thrown if not for sorting. 

Then I realized that maybe relying on ordering received from the converter is the correct choice respecting separation of responsibilities of system's layers.

In the end I think that we should delay the choice to compatibility test with the Confluent's sink and only then decide what to do.
What do you think?

brandtnewton replied:
Good point, let's wait to see how the compatibility tests look. If we decide we want to sort the map, we might be able to reuse one of your value to byte[] serializers to serialize the keys and then sort the byte arrays.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/BigtableSinkTaskTest.java:
```diff
@@ -0,0 +1,612 @@
+
+  @Test
+  public void testGetTableName() {
+    SinkRecord record = new SinkRecord("topic${test}", 1, null, null, null, null, 1);
```
It's not an edge case and the test was to prove it. It was a wrong choice - I removed `"${test}"`.
Now `noSubstitution` is a test case with no interpolations characters.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/BigtableSinkTaskTest.java:
```diff
@@ -0,0 +1,612 @@
+
+    byte[] rowKey = "rowKey".getBytes(StandardCharsets.UTF_8);
+    doReturn(rowKey).when(keyMapper).getKey(any());
+    doAnswer(
```
As you can see, it was probably simpler than writing the mocks in this case.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/autocreate/BigtableSchemaManagerTest.java:
```diff
@@ -0,0 +1,749 @@
+  }
+
+  @Test
+  public void testErrorsCreatingColumnFamilies() {}
```
Removed (it's tested in different cases: `testEnsureColumnFamiliesExistSomeSomeErrorsDueToConcurrentColumnFamilyDeletion`, `testEnsureColumnFamiliesExistSomeCreatedSuccessfullySomeErrorsDueToRacesOrInvalidRequests`).
Sorry, I didn't mark it with "TODO" so I missed it when cleaning up the code.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/autocreate/SchemaApiExceptionsTest.java:
```diff
@@ -0,0 +1,143 @@
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class SchemaApiExceptionsTest {
```
Done.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapperTest.java:
```diff
@@ -0,0 +1,835 @@
+            calculateKey(List.of(fieldName), DELIMITER, kafkaConnectStruct),
+            fieldByteValue.toString().getBytes(StandardCharsets.UTF_8)));
+
+    final Short fieldShortValue = (short) 4242;
```
Done.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapperTest.java:
```diff
@@ -0,0 +1,835 @@
+            calculateKey(List.of(fieldName), DELIMITER, kafkaConnectStruct),
+            fieldFloatValue.toString().getBytes(StandardCharsets.UTF_8)));
+
+    final Double fieldDoubleValue = 4242424242.4242;
```
Done.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/mapping/KeyMapperTest.java:
```diff
@@ -0,0 +1,835 @@
+            calculateKey(List.of(innerFieldIntegerName), DELIMITER, kafkaConnectInnerStruct),
+            innerIntegerValue.toString().getBytes(StandardCharsets.UTF_8)));
+
+    Schema kafkaConnectMiddleSchema =
```
It'd need a lot of copying since nested structs are used widely here. For example `kafkaConnectInnerStruct` is used thorough the whole test case.
Maybe I should restructure this test function instead? E.g., first schemas, then structs, then organized assertions with comments specifying what's being tested?

brandtnewton replied:
I think it's fine to repeat non-logic code in tests. Can you give splitting the test up a try? If it ends up getting messy you can bail on it 😄

prawilny replied:
Done.

prawilny commented:
@brandtnewton
I didn't resolve any of the conversations since I'm used to the reviewer doing that.

I also have a question: how do you want to review further commits (mainly logical types support, some more integration tests, and some minor tweaks throughout the codebase)? In this PR? In a new one? Or maybe do you want to create a new repository and have the PR(s) sent there? Please let me know.

brandtnewton reviewed (changes requested):
Thanks for all the changes! Feel free to add more changes here or in another PR, whatever is easier for you. See comments:

* [try splitting the test up](https://github.com/googleapis/java-bigtable/pull/2466#discussion_r1934231952)
* [waiting to see how the compatibility tests look](https://github.com/googleapis/java-bigtable/pull/2466#discussion_r1933989270)

prawilny committed:
split testStruct into three

prawilny committed:
update pom.xml with integration test dependencies

prawilny committed:
use config variable in JsonConverterFactory

prawilny committed:
Unify quoting variables in log messages

prawilny committed:
Make column family presence required for delete operations

prawilny committed:
Add string interpolation to default column family config

prawilny committed:
Support logical types

prawilny committed:
More integration tests

prawilny commented:
@brandtnewton, please take a look.
Further integration tests will follow (e.g., more comprehensive error handling).
I will also document the generation of data for compatibility tests, but I'm delaying it to only document the final version of it which hopefully will be able to make use of the confluent sink with the bugs in logical types support fixed.

prawilny committed:
Provide service provider configuration file

prawilny committed:
Move DLQ helpers to a base class

prawilny committed:
Error handling Integration Tests

prawilny committed:
Sort input records for predictable batching behavior

prawilny committed:
Error handling and resource auto creation integration tests (and some integration test refactors)

prawilny committed:
Unbreak DifferentConvertersIT

prawilny committed:
BigtableSinkConfig: use "_CONFIG" as a suffix of static variables denoting configuration keys for consistency with Kafka convention

prawilny committed:
Configure Admin API write retries

prawilny committed:
Configure integration test logging

prawilny committed:
Trace level logs for resource creation

prawilny committed:
Stop considering FAILED_PRECONDITION an input error

prawilny committed:
Use Parameterized for DifferentConvertersIT

prawilny committed:
Integration test cleanup

move code from BaseIT to BaseKafkaConnectIT

get rid of protected and use public/private in base classes

prawilny committed:
Warning about VersionIT

prawilny committed:
Tweak number of partitions for created topics

prawilny committed:
Integration test using multiple connector tasks

prawilny committed:
Remove a disabled test that nukes request quota

brandtnewton commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/integration/ResourceAutoCreationIT.java:
```diff
@@ -0,0 +1,393 @@
+  public void testCreationOfInvalidTable() throws InterruptedException {
+    String dlqTopic = createDlq();
+    Map<String, String> props = baseConnectorProps();
+    String invalidTableName = "T".repeat(100);
```
nit: MAX_BIGTABLE_TABLE_NAME_LENGTH + 1 instead of `100` to make it more clear why this is invalid

prawilny replied:
Done.

brandtnewton commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/integration/ResourceAutoCreationIT.java:
```diff
@@ -0,0 +1,393 @@
+  }
+
+  @Test
+  public void testDeletionFailsWhenAutoCreationDisabled() throws InterruptedException {
```
is this a desired behavior?

prawilny replied:
It's a test documenting consequences of design choices described in comments in `MutationDataBuilder` (in `deleteFamily()` and `deleteCells()`), so I think it is a desired behavior. The alternative is described in these comments and is - in my opinion - worse.
 
I added a comment to make that connection explicit.

brandtnewton reviewed (commented):
I think I'm all caught up now, including the tests. This is getting super close! I'd like to run the IT tests locally, can you add a quick doc on how to do that? Also I don't have permission to resolve comments so feel free to resolve them yourself.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/mapping/MutationDataBuilder.java:
```diff
@@ -63,12 +63,30 @@ public void deleteRow() {
+    // b) modify the record as if it didn't contain the deletion of columns that don't exist so that
+    // other operations within the row have a chance to execute successfully.
+    // I think the option a) is clearer to reason about, so we mark the column family as required.
+    requiredColumnFamilies.add(columnFamily);
```
As per the design doc:
"""
The SinkRecord value has a top level field set to null
Example: {foo: null}. This will result in all columns in the "foo" column family being deleted for this row.
"""

We create a column family in the table's schema and then delete all the cells within the row.
Yes, the "global" (within a table) creation is required only so that the Bigtable modification request succeeds, but that modification is not reverted by a column family deletion within a row.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/BigtableSinkTask.java:
```diff
@@ -301,6 +309,32 @@ void reportError(SinkRecord record, Throwable throwable) {
+   */
+  @VisibleForTesting
+  // It is generic so that we can test it with naturally ordered values easily.
+  static <K, V> Map<K, V> orderMap(Map<K, V> map, Collection<K> order) {
```
It could, but then we'd need to also return a LinkedHashMap from `autoCreateTablesAndHandleErrors()` and `autoCreateColumnFamiliesAndHandleErrors()`. I thought that it'd couple these functions too tightly. What do you think?

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/autocreate/BigtableSchemaManager.java:
```diff
@@ -461,7 +468,6 @@ static Optional<StatusCode> maybeExtractBigtableStatusCode(Throwable t) {
     static boolean isStatusCodeCausedByInputError(StatusCode.Code code) {
       switch (code) {
         case INVALID_ARGUMENT:
-        case FAILED_PRECONDITION:
```
The [docs](https://cloud.google.com/bigtable/docs/status-codes#returned_codes) state that it is a "user error", but also that "Depending on the circumstance, you might want to retry. For example, you can encounter this error if you attempt to perform operations on a table while it is still being created.".

I encountered this status code when creating different column families concurrently. The error message directly stated that it was due to concurrent table schema modification. That is why it needs to be retried in our case.

I also think that the other example error given by the docs, "attempt to perform operations on a table while it is still being created", should also be retried if it were encountered (due to either a bug in our resource auto creation or the user's too late manual creation of the table if the sink were used with the resource auto creation disabled).

Apart from that, I don't really know if a user input error *could* cause a failed precondition error - I didn't  manage to find a direct answer in the docs. I only know that the invalid operations I test (such as invalid table name) cause INVALID_ARGUMENT and at least some of the FAILED_PRECONDITION errors should be retried.

Do you think we should read error messages and decide whether the exception is caused by user error depending on that? If so, do you have any pointers where to find a more or less complete list of the error messages?

prawilny replied:
Also, I think we should err on leniency's side with branding something a user error since in that case (if the user configured DLQ) we send the offending record to the DLQ with InvalidBigtableSchemaModificationException, which we describe in the docs as:
```java
/**
 * An {@link Exception} that signifies that input {@link org.apache.kafka.connect.sink.SinkRecord
 * SinkRecord(s)} cause attempt of invalid Cloud Bigtable schema modification and thus is invalid
 * and should not be retried.
 */
```

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/config/BigtableSinkConfig.java:
```diff
@@ -366,46 +398,67 @@ private static void addErrorMessage(
+        .modifyColumnFamiliesSettings()
+        .setRetrySettings(adminApiWriteRetrySettings)
+        // Retry createTable() for status codes other admin operations retry by default as
+        // seen in BigtableTableAdminStubSettings and for FAILED_PRECONDITION which is
```
No, then the caller receives `ALREADY_EXISTS` error, which is not retriable. Then `awaitResourceCreationAndHandleInvalidInputErrors()` logs that information and then swallows the exception, then `refreshTableColumnFamiliesCache()` gets the table's schema and the returned table's schema is used to check if the column family creation succeeded.

prawilny committed:
Use SchemaAndValue in KeyMapper#extractField().

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/config/BigtableSinkConfig.java:
```diff
@@ -314,33 +342,37 @@ public static ConfigDef getDefinition() {
+            DEFAULT_COLUMN_FAMILY_CONFIG,
             ConfigDef.Type.STRING,
-            "default",
+            ConfigInterpolation.TOPIC_PLACEHOLDER,
```
Unfortunately, the current implementation is how the Confluent's sink behaves.
Should I change it to `"default"` and break compatibility?

prawilny committed:
Implement hashCode() for SchemaUtils.SchemaAndField

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/mapping/SchemaUtils.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
If I recall my reading the docs when writing this class, strictly speaking we don't have to. But I agree, we should implement it, my omission of it was a lapse of judgement.
Done.

prawilny committed:
Assert against case insensivity of enums in BigtableSinkConfigTest#testEnumCaseInsensitivity

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/config/BigtableSinkConfigTest.java:
```diff
@@ -81,47 +82,58 @@ public void testBasicValidationFailure() {
-    props.put(CONFIG_VALUE_NULL_MODE, "delete");
+    props.put(INSERT_MODE_CONFIG, "uPsErT");
+    props.put(ERROR_MODE_CONFIG, "IGNORE");
+    props.put(VALUE_NULL_MODE_CONFIG, "delete");
```
Done.

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/config/BigtableSinkConfigTest.java:
```diff
@@ -81,47 +82,58 @@ public void testBasicValidationFailure() {
+    props.put(INSERT_MODE_CONFIG, "uPsErT");
+    props.put(ERROR_MODE_CONFIG, "IGNORE");
+    props.put(VALUE_NULL_MODE_CONFIG, "delete");
     BigtableSinkConfig config = new BigtableSinkConfig(props);
```
Done.

prawilny committed:
Fix typo in 'getStructhWithNullOnNthNestingLevel'

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/util/NestedNullStructFactory.java:
```diff
@@ -0,0 +1,49 @@
+  public static final ByteString NESTED_NULL_STRUCT_FIELD_NAME_BYTES =
+      ByteString.copyFrom(NESTED_NULL_STRUCT_FIELD_NAME.getBytes(StandardCharsets.UTF_8));
+
+  public static Struct getStructhWithNullOnNthNestingLevel(int n) {
```
Done.

prawilny committed:
State explicitly why the operations in ErrorReportingIT fail

prawilny committed:
Improve InsertUpsertIT: check DLQ, assert against values

prawilny commented on google-cloud-bigtable-kafka-connect-sink/src/test/java/com/google/cloud/kafka/connect/bigtable/integration/InsertUpsertIT.java:
```diff
@@ -0,0 +1,86 @@
+    props.put(BigtableSinkConfig.AUTO_CREATE_TABLES_CONFIG, "true");
+    props.put(BigtableSinkConfig.AUTO_CREATE_COLUMN_FAMILIES_CONFIG, "true");
+    props.put(BigtableSinkConfig.INSERT_MODE_CONFIG, InsertMode.INSERT.name());
+    props.put(BigtableSinkConfig.ERROR_MODE_CONFIG, BigtableErrorMode.IGNORE.name());
```
Done.

prawilny committed:
Use MAX_BIGTABLE_TABLE_NAME_LENGTH in ResourceAutoCreationIT#testCreationOfInvalidTable()

prawilny committed:
Explain why testDeletionFailsWhenAutoCreationDisabled is needed

prawilny committed:
Readme for tests

prawilny commented:
@brandtnewton, please take another look.
I responded to all your comments, resolved the conversations that were done (but I left [the one you didn't mark as read](https://github.com/googleapis/java-bigtable/pull/2466#discussion_r1925587703), thinking that you might want to take a look at it since it still seems relevant).
I also created a small README about running integration tests.

Beware, the integration tests are a bit flaky right now.
I'm going to fix it right after preparing the performance tests.
The problematic code path is: auto creation of a table => auto creation of a column family => access to that created column family. Sometimes that access receives NOT_FOUND status code from its request and the test fails.

brandtnewton reviewed (commented):
Great doc. Thanks for adding it!

prawilny committed:
Copy OpenTelemetry collector deployment from https://github.com/GoogleCloudPlatform/otlp-k8s-ingest/tree/3b383a083250487e5ca27e58abf0d81cb25a914c/k8s/base

prawilny committed:
Use Confluent's sink in compatibility tests directly through a .jar

prawilny committed:
Warn against latency unpredictability when using table and/or column family auto creation

prawilny committed:
Download and normalize strimzi 0.45.0 with namespace 'kafka' configured

curl "https://strimzi.io/install/latest?namespace=kafka" | yq -P 'sort_keys(..)' | sponge strimzi-cluster-operator-0.45.0.yaml

prawilny committed:
Copy Strimzi Kafka Connect metrics config from https://github.com/strimzi/strimzi-kafka-operator/blob/release-0.45.x/examples/metrics/kafka-connect-metrics.yaml

prawilny committed:
Bump Google libraries

prawilny committed:
Performance tests

prawilny committed:
MultipleConnectorTasksIT#testRestartPauseStop

prawilny committed:
Remove at-least assertions with exact ones

prawilny commented:
@brandtnewton, please take another look.
The chagnes are mostly with the performance test setup and compatibility tests, but a few lines (documentation strings) are also changed in the library code.

prawilny committed:
Configure retries for CheckAndMutate

prawilny committed:
Update READMEs

prawilny committed:
Use required dependency versions

prawilny committed:
Include integration test downloadables in .gitgnore

prawilny committed:
Use getName() for getting class name everywhere

prawilny committed:
Use file-based credentials for all integration tests to avoid special casing ConfluentCompatibilityIT

prawilny committed:
Handle temporary errors in integration test helpers

prawilny committed:
Fix comment about retries in Bigtable Admin API

prawilny committed:
Test that ServiceLoader works as expected

prawilny committed:
Get rid of test flakiness by creating the Bigtable tables and column families before the integration tests

prawilny committed:
Avoid using deprecated methods

prawilny committed:
Upgrade JUnit to 4.13.2

prawilny committed:
Get rid of test flakiness by waiting for the sink-induced schema modifications before starting to read table rows

prawilny committed:
Configure workaround for NOT_FOUND and FAILED_PRECONDITION after resource auto creation

prawilny committed:
Move jar assembly config into a subdirectory

prawilny committed:
Separate sink and its integration tests into spearate maven submodules

prawilny committed:
Provide a simple README describing code organization and linking to the other READMEs

prawilny commented:
@brandtnewton, please take a look.
Just about everything is done now. The main missing thing seems to be the CI config, but it waits for the creation of the new repository.

brandtnewton reviewed (commented):
There might be some junk files in here but otherwise lgtm

brandtnewton reviewed (changes requested):
Please add this extra header so we can monitor usage. That goes in ~line 538

```
import com.google.api.gax.rpc.FixedHeaderProvider;
...
    dataStubSettings.setHeaderProvider(
        FixedHeaderProvider.create("user-agent", "bigtable-java/kafka-connect-sink/" + PackageMetadata.getVersion()));
```

prawilny committed:
Set headers on requests to Bigtable API

prawilny commented on google-cloud-bigtable-kafka-connect-sink/doc/performance/enable_tracing.patch:
```diff
@@ -0,0 +1,69 @@
+diff --git a/google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/BigtableSinkTask.java b/google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/BigtableSinkTask.java
+index 80bea508..3dcf2b27 100644
+--- a/google-cloud-bigtable-kafka-connect-sink/src/main/java/com/google/cloud/kafka/connect/bigtable/BigtableSinkTask.java
```
Yes, see this section in the performance tests' README (`google-cloud-bigtable-kafka-connect-sink/doc/performance/README.md`):
```md
### Tracing code
Tracing code is not present in the release code.
The code of tracing implementation is available in [`MessageTracer.java`](MessageTracer.java).
To enable tracing, put it into `sink/src/main/java/com/google/cloud/kafka/connect/bigtable/tracing` and apply [`enable_tracing.patch` git patch](enable_tracing.patch).
```

I think it is a better idea to show the diff rather than store the whole file since it can be more easily applied on changed file in the future to rerun the tests, but if you prefer to do it in some other way, I'll change it.

prawilny committed:
Remove an unused file

prawilny commented:
Header added to both Admin and Data clients, hopefully it wasn't overzealous.

In one of your previous comments you mentioned "some junk files".
Let me explain the categories of not-strictly-necessary files, so that you might more easily point to the argument you disagree with so that I can remove the junk.

There are symlink present in `integration-tests` directory.
Some of them (`log4j.properties`, `fake_service_key.json`) are just resources.
Some of them (the config enums, *Config classes, `PackageMetadata`, the exception(s)) contain strings we want to use in the integration tests (be it "gcp.bigtable.project.id" or "com.google.cloud.kafka.connect.bigtable.exception.InvalidBigtableSchemaModificationException").
Some of them (`JsonConverterFactory`, `NestedNullStructFactory`, client constructors in `BigtableSinkConfig`) are helpers whose code is needed by both the integration tests and either release code or the unit tests.
The alternatives I came up with were:
- the current symlink solution,
    - might make it impossible to run the integration tests on Windows,
    - does not touch the release code,
- copypaste of the code,
    - changes to anything need to be manually propagated,
- some additional maven module containing all the common code that both `sink` and `integration-test` modules would depend on,
    - complicates maven structure,
- some kind of generation of the strings in a single place and populating the source files during the compilation
    - does not solve the problem with helpers,
    - risky since it touches the release code,
I think the symlinks are a surprisingly clean solution to this problem - I think we don't really care about the integration tests on Windows and the other solutions are more complicated in my opinion.
What do you think?

There are k8s manifests present in `doc/performance` directory.
We could possibly put them elsewhere, but I couldn't come up with anything better than the repo.
Do you have a better idea?

I think all the other (`git ls-files | grep -v \.java`) files are either:
- code license, READMEs, `.gitignore`s and other well-known files
- maven (and its plugins') configs (`pom.xml`s, `sink/src/main/assembly/package.xml`, `license.header`)
- performance test documentation (everything in `doc/performance`)
- test resources (in `*/src/test/resources`)
    - I rechecked in `ConfluentCompatibilityIT.java` that all the files from `integration-tests/src/test/resources/compatibility_test_cases` are used`
- Java ServiceLoader provider file (`sink/src/main/resources/META-INF/services/org.apache.kafka.connect.sink.SinkConnector`)
- the symlinks described above
    - I rechecked that all the symlinks are needed now (in process I found a single unused file, which I deleted).

So can I get an approval of this PR without any caveats or do I still need to do something? Apart from the Kokoro config, but it's going to happen only in the new repo, right?

brandtnewton reviewed (commented):
LGTM!
